corgi/collectors/models.py

```diff
@@ -43,21 +43,45 @@ def __str__(self):
         return f"{self.nvr}"
 
 
-class CollectorComposeSRPM(models.Model):
+class CollectorSRPM(models.Model):
 
     build_id = models.IntegerField(primary_key=True)
 
     def __str__(self):
         return f"{self.build_id}"
 
 
-class CollectorComposeRPM(models.Model):
+class CollectorRPM(models.Model):
     """Not every RPM built as part of a SRPM is included in a module
     Which is why we don't directly related SRPM builds to product_streams using relations table"""
 
     nvra = models.TextField(unique=True)
-    rhel_module = models.ManyToManyField(CollectorComposeRhelModule)
-    srpm = models.ForeignKey(CollectorComposeSRPM, on_delete=models.CASCADE, related_name="rpms")
+    rhel_module = models.ManyToManyField(CollectorRhelModule)
+    srpm = models.ForeignKey(CollectorSRPM, on_delete=models.CASCADE, related_name="rpms")
 
     def __str__(self):
         return f"{self.nvra}"
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                name="unique_collectorrpms",
+                fields=("nvra", "srpm"),
+            ),
+        ]
```

Comment:
I think this isn't needed since we already require NVRA to be unique by itself.

Reply:
This was a suggestion from @JimFuller-RedHat 
I reached out to him for help because using get_or_create with an nvra for CollectorRPM model manager was throwing duplicate key errors, see [this comment](https://issues.redhat.com/browse/CORGI-163?focusedCommentId=20988378&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20988378) on the CORGI-163 task. I confirmed this suggestion of a UniqueConstraint fixed that issue.

Reply:
Hmm, I don't get how a redundant UniqueConstraint would fix that issue. Maybe get_or_create() wasn't behaving atomically because we didn't have an index for both its keyword arguments (nvra and srpm), so it was trying to insert duplicate rows and failing on the unique nvra requirement.

In that case you should be able to just explicitly add an index on both nvra and srpm, instead of relying on UniqueConstraint (which implicitly creates an index on all the fields in the constraint IIUC). As far as I'm aware, though, if you have a unique constraint / index on a single field, get_or_create() should be atomic - you don't need the constraint / index on all fields.

Either way I'd like to have a better idea of why this bug happens before we merge this code. I don't want us to assume it's fixed, only to see it happen again later. For example, if it was triggered by some kind of race condition which just happened to not occur after the constraint changes, but which is still present in the code.

Reply:
Other then the general - 'sometimes ORMs does surprising things with sql' (thankfully rare with django)... I do not have any good explanation - as this is a collector model I think the risk is small - having no free cycles to dig into a root cause ... might be worth raising a task onto the backlog if/when this does arise in the future.

Reply:
Rereading CORGI-163, just the unique=True constraint on nvra should be enough to prevent duplicates from being inserted. There's really no need for a separate UniqueConstraint on both nvra + srpm together.

I assume when you saw the errors with duplicate keys, you added both the unique=True condition on nvra and the UniqueConstraint on (nvra, srpm) as mentioned in Jim's comment. They're both doing the same job and only one of them is needed - let's either force nvra to be unique by itself, or force nvra + srpm to be unique together.

Reply:
I removed unique=True from the nvra field and kept the uniqueconstraint on nvra + srpm. I also have to updated the CollectorRPM.get_or_create() call in Brew().persist_modules to include srpm in the fields instead of in default vaules.